File: scripts/test-phase5-integration.mjs
/**
 * LeadPulse AI V2 — Phase 5 End-to-End Integration Verification Script
 *
 * Proves the complete pipeline:
 * Next.js Environment -> Python Worker -> Lead Triage Agent
 *   -> Parallel Specialist Agents (Website, Ads, Maps)
 *   -> Strict Anti-Hallucination Guardrails & Isolated Fault Tolerance
 */

import { spawn } from 'child_process';
import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';
import dotenv from 'dotenv';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const projectRoot = resolve(__dirname, '..');

// 1. Load project environment variables
dotenv.config({ path: resolve(projectRoot, '.env') });
dotenv.config({ path: resolve(projectRoot, '.env.local') });

const workerUrl = (process.env.AI_WORKER_URL || 'http://127.0.0.1:8000').replace(/\/+$/, '');
const providerUrl = process.env.AI_BASE_URL;
const apiKey = process.env.AI_API_KEY;
const model = process.env.AI_MODEL || 'auto';

console.log('='.repeat(70));
console.log('LeadPulse AI V2 — Phase 5 Integration Verification');
console.log('='.repeat(70));
console.log(`Project Environment Root: ${projectRoot}`);
console.log(`Target Worker URL:        ${workerUrl}`);
console.log(`Target Provider Base:     ${providerUrl}`);
console.log(`Configured Model:         ${model}`);
console.log(`AI API Key Status:        ${apiKey ? '[CONFIGURED - MASKED]' : '[MISSING]'}`);
console.log('='.repeat(70));

if (!providerUrl || !apiKey) {
  console.error('FATAL: Missing required environment variables (AI_BASE_URL or AI_API_KEY).');
  process.exit(1);
}

let workerProcess = null;

async function probeWorkerHealth() {
  try {
    const res = await fetch(`${workerUrl}/health`);
    if (res.ok) {
      const data = await res.json();
      return data.status === 'healthy';
    }
  } catch {
    return false;
  }
  return false;
}

async function startWorkerIfNeeded() {
  const isRunning = await probeWorkerHealth();
  if (isRunning) {
    console.log(`[Worker] Detected active worker at ${workerUrl}. Reusing running instance.`);
    return;
  }

  console.log(`[Worker] No active worker detected. Spawning Uvicorn worker process on ${workerUrl}...`);
  const pythonCmd = process.platform === 'win32' ? 'python' : 'python3';
  workerProcess = spawn(
    pythonCmd,
    ['-m', 'uvicorn', 'app.main:app', '--app-dir', 'ai-worker', '--host', '127.0.0.1', '--port', '8000'],
    {
      cwd: projectRoot,
      env: { ...process.env, AI_WORKER_HOST: '127.0.0.1', AI_WORKER_PORT: '8000' },
      stdio: ['ignore', 'pipe', 'pipe'],
    }
  );

  workerProcess.stdout.on('data', (d) => {
    const line = d.toString().trim();
    if (line) console.log(`[Worker STDOUT] ${line}`);
  });

  workerProcess.stderr.on('data', (d) => {
    const line = d.toString().trim();
    if (line && !line.includes('GET /health')) console.log(`[Worker STDERR] ${line}`);
  });

  let attempts = 0;
  while (attempts < 30) {
    await new Promise((r) => setTimeout(r, 500));
    if (await probeWorkerHealth()) {
      console.log('[Worker] Python AI worker started and responding healthy.');
      return;
    }
    attempts++;
  }

  throw new Error('Timed out waiting for Python AI worker to become ready.');
}

function stopWorker() {
  if (workerProcess) {
    console.log('[Worker] Shutting down spawned worker process...');
    workerProcess.kill('SIGTERM');
    workerProcess = null;
  }
}

async function runStep(stepNumber, description, fn) {
  console.log(`\n[STEP ${stepNumber}] ${description}`);
  try {
    await fn();
    console.log(`✓ [STEP ${stepNumber} PASSED]`);
  } catch (err) {
    console.error(`✗ [STEP ${stepNumber} FAILED]:`, err.message);
    stopWorker();
    process.exit(1);
  }
}

async function main() {
  try {
    await startWorkerIfNeeded();

    // Step 1: Health & Readiness
    await runStep(1, 'Verify Worker Health and System Readiness', async () => {
      const res = await fetch(`${workerUrl}/health/ready`);
      if (!res.ok) throw new Error(`Readiness check returned HTTP ${res.status}`);
      const data = await res.json();
      if (!data.provider_accessible) throw new Error('Worker reported provider_accessible=false');
      console.log(`  Readiness: status=${data.status}, accessible=${data.provider_accessible}`);
    });

    // Step 2: Live Lead Triage Agent Execution
    await runStep(2, 'Invoke Lead Triage Agent to Evaluate Prospective Candidate', async () => {
      const prompt =
        'Evaluate prospective business candidate "Aura Dental Lounge" located in Gurgaon:\n' +
        'Business Type: Aesthetic Dentistry\n' +
        'Scorecard Score: 9/10\n' +
        'Active Ads: True (Running teeth whitening & veneer campaigns)\n' +
        'Website: https://auradental.in\n' +
        'Google Rating: 4.8 across 95 reviews\n' +
        'Friction: Slow response to website WhatsApp inquiries\n' +
        'Return a structured LeadTriageResult.';

      const res = await fetch(`${workerUrl}/api/v1/run`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          prompt,
          agent_type: 'lead_triage',
          timeout_seconds: 120,
        }),
      });

      if (!res.ok) {
        const text = await res.text();
        throw new Error(`Execution returned HTTP ${res.status}: ${text}`);
      }

      const data = await res.json();
      if (!data.success) throw new Error(`Agent execution failed: ${data.error}`);
      const out = data.output;
      if (typeof out.qualified !== 'boolean' || !out.priority || !out.reason || !Array.isArray(out.research_agents)) {
        throw new Error('Output does not conform to LeadTriageResult schema');
      }

      console.log(`  Qualified: ${out.qualified} | Priority: ${out.priority}`);
      console.log(`  Reason: "${out.reason}"`);
      console.log(`  Selected Specialists: [${out.research_agents.join(', ')}]`);
    });

    // Step 3: End-to-End Parallel Specialist Research
    await runStep(3, 'Execute Parallel Specialist Research Endpoint with Failure Isolation', async () => {
      const candidatePayload = {
        business_name: 'Aura Dental Lounge',
        business_type: 'Aesthetic Dentistry',
        source_platform: 'google_maps',
        source_url: 'https://maps.google.com/auradental',
        scorecard_score: 9,
        priority_tier: 'immediate',
        website_url: 'https://auradental.in',
        has_active_ads: true,
        rating: 4.8,
        review_count: 95,
        address: 'Golf Course Road, Gurgaon',
        phone_number: '+919876543210',
        raw_metadata: {
          category: 'Dental clinic',
          headline: 'Veneers & Smile Makeovers 20% Off',
          ad_creative_text: 'Transform your smile in 2 visits. Book free 3D smile scan.',
          cta_text: 'Send WhatsApp Message',
          ad_count: 1,
        },
      };

      const preTriage = {
        qualified: true,
        priority: 'high',
        reason: 'High-ticket cosmetic clinic with active ads',
        research_agents: ['ads', 'maps'],
      };

      const res = await fetch(`${workerUrl}/api/v1/specialists/research`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          candidate: candidatePayload,
          triage_result: preTriage,
          limits: { max_deep_research_leads: 5 },
        }),
      });

      if (!res.ok) {
        const text = await res.text();
        throw new Error(`Execution returned HTTP ${res.status}: ${text}`);
      }

      const aggregate = await res.json();
      if (!aggregate.business_name || !aggregate.triage_result || !aggregate.execution_status) {
        throw new Error('Aggregate output missing required structure');
      }

      console.log(`  Researched Business: ${aggregate.business_name}`);
      console.log(`  Execution Status: ${JSON.stringify(aggregate.execution_status)}`);
      console.log(`  Total Wall Time: ${aggregate.execution_time_ms} ms`);
      console.log(`  Token Usage: ${JSON.stringify(aggregate.token_usage)}`);

      if (aggregate.ads_analysis) {
        console.log(`  Ads Status: ${aggregate.ads_analysis.status} (Active Creatives: ${aggregate.ads_analysis.active_ad_count})`);
      }
      if (aggregate.maps_analysis) {
        console.log(`  Maps Status: ${aggregate.maps_analysis.status} (Rating: ${aggregate.maps_analysis.rating}, Sentiment: ${aggregate.maps_analysis.review_sentiment || 'unknown'})`);
      }
    });

    // Step 4: Credential Non-Exposure Check
    await runStep(4, 'Verify Zero Credential Leakage in Output Payloads', async () => {
      const res = await fetch(`${workerUrl}/api/v1/run`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          prompt: 'Evaluate clinic "Test Dental".',
          agent_type: 'lead_triage',
          timeout_seconds: 60,
        }),
      });

      const bodyText = await res.text();
      if (!res.ok) throw new Error(`Execution returned HTTP ${res.status}: ${bodyText}`);
      if (bodyText.includes(apiKey)) {
        throw new Error('FATAL SECURITY LEAK: AI_API_KEY detected in response payload!');
      }
      console.log('  Confirmed zero secret key leakage across response body.');
    });

    console.log('\n' + '='.repeat(70));
    console.log('ALL PHASE 5 INTEGRATION CHECKS PASSED SUCCESSFULLY');
    console.log('='.repeat(70));
  } finally {
    stopWorker();
  }
}

main().catch((err) => {
  console.error('Fatal unexpected error:', err);
  stopWorker();
  process.exit(1);
});

File: tests/test_phase5_integration.py
"""LeadPulse AI V2 — Phase 5 Live Integration Test Suite.

Proves the complete end-to-end integration:
NormalizedCandidate
    │
    ▼
Lead Triage Agent (Live Upstream LLM Provider via OpenAI Agents SDK)
    │
    ├──── Website Specialist Agent
    │
    ├──── Ads Specialist Agent
    │
    └──── Maps Specialist Agent
             │
             ▼
SpecialistResearchAggregate (Aggregated evidence, failure isolation, strict anti-hallucination)
"""

import os
import sys
import unittest
from pathlib import Path
from dotenv import load_dotenv
import httpx
from httpx import ASGITransport, AsyncClient

# Add project root and ai-worker to sys.path
_PROJECT_ROOT = Path(__file__).resolve().parent.parent
_AI_WORKER_DIR = _PROJECT_ROOT / "ai-worker"
if str(_PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(_PROJECT_ROOT))
if str(_AI_WORKER_DIR) not in sys.path:
    sys.path.insert(0, str(_AI_WORKER_DIR))

from app.core.config import get_settings
from app.core.provider import close_provider, init_provider
from app.main import app
from app.schemas.profile import CampaignScope
from app.schemas.search import CampaignLimits
from app.schemas.evidence import (
    Evidence,
    LeadTriageResult,
    WebsiteAnalysisResult,
    AdsAnalysisResult,
    MapsAnalysisResult,
    SpecialistResearchAggregate,
)
from app.services.deterministic_pipeline import NormalizedCandidate
from app.services.specialist_orchestrator import SpecialistOrchestratorService


class TestPhase5LiveIntegration(unittest.IsolatedAsyncioTestCase):
    """Live end-to-end integration test suite verifying Phase 5 Lead Triage and Specialist Agents."""

    async def asyncSetUp(self):
        load_dotenv(dotenv_path=_PROJECT_ROOT / ".env")
        self.settings = get_settings()
        init_provider(self.settings)
        self.transport = ASGITransport(app=app)
        self.client = AsyncClient(transport=self.transport, base_url="http://localhost:8000")

    async def asyncTearDown(self):
        await self.client.aclose()
        await close_provider()

    async def test_01_lead_triage_agent_live_execution(self):
        """Verify Lead Triage Agent evaluates a candidate and returns structured LeadTriageResult."""
        prompt = (
            "Evaluate prospective business candidate 'Apex Dental Spa' located in Gurgaon:\n"
            "Business Type: Cosmetic Dentistry & Dental Implants\n"
            "Scorecard Score: 9/10 (High ticket, verified location)\n"
            "Active Ads: True (Running Invisalign and implant offers)\n"
            "Website: https://apexdentalspa.in\n"
            "Google Rating: 4.9 across 140 reviews\n"
            "Friction Points: Manual WhatsApp booking, slow intake response\n"
            "Return a structured LeadTriageResult specifying qualification, priority, reason, and research_agents."
        )

        response = await self.client.post(
            "/api/v1/run",
            json={
                "prompt": prompt,
                "agent_type": "lead_triage",
                "timeout_seconds": 120.0
            }
        )
        self.assertEqual(response.status_code, 200)
        data = response.json()
        self.assertTrue(data["success"], f"Lead triage agent execution failed: {data.get('error')}")
        self.assertEqual(data["agent_type"], "lead_triage")

        output = data["output"]
        self.assertIsInstance(output, dict)
        self.assertIn("qualified", output)
        self.assertIn("priority", output)
        self.assertIn("reason", output)
        self.assertIn("research_agents", output)

        validated = LeadTriageResult.model_validate(output)
        self.assertTrue(validated.qualified)
        self.assertIn(validated.priority, ["high", "medium", "low"])
        self.assertGreater(len(validated.research_agents), 0)

    async def test_02_website_specialist_live_execution(self):
        """Verify Website Specialist analyzes observable signals and produces structured WebsiteAnalysisResult."""
        prompt = (
            "Analyze the following public website inspection data for 'Apex Dental Spa':\n"
            "URL: https://apexdentalspa.in\n"
            "Title: Apex Dental Spa | Premier Cosmetic & Implant Clinic\n"
            "Meta Description: Best cosmetic dentist in Gurgaon offering smile makeovers.\n"
            "Text Summary: Welcome to Apex Dental Spa. We specialize in dental implants and smile design. "
            "Contact our team today to schedule your consultation.\n"
            "Observed CTAs: ['Book Free Consultation', 'Call +91 9876543210', 'Chat on WhatsApp']\n"
            "WhatsApp Links: ['https://wa.me/919876543210']\n"
            "Booking Links: ['https://calendly.com/apexdental/consult']\n"
            "Has Booking System: True\n"
            "Has WhatsApp CTA: True\n"
            "Return a structured WebsiteAnalysisResult."
        )

        response = await self.client.post(
            "/api/v1/run",
            json={
                "prompt": prompt,
                "agent_type": "website_specialist",
                "timeout_seconds": 120.0
            }
        )
        self.assertEqual(response.status_code, 200)
        data = response.json()
        self.assertTrue(data["success"], f"Website specialist failed: {data.get('error')}")

        output = data["output"]
        validated = WebsiteAnalysisResult.model_validate(output)
        self.assertIn(validated.status, ["available", "partial", "unavailable"])
        self.assertGreaterEqual(validated.confidence, 0.0)
        self.assertLessEqual(validated.confidence, 1.0)
        self.assertIsInstance(validated.evidence, list)

    async def test_03_ads_specialist_live_execution_no_metric_fabrication(self):
        """Verify Ads Specialist analyzes Meta Ads copy and does NOT fabricate spend or ROAS."""
        prompt = (
            "Analyze Meta Ads data for 'Apex Dental Spa':\n"
            "Business Name: Apex Dental Spa\n"
            "Has Active Ads: True\n"
            "Ad Headline: Get 30% Off Dental Implants This Month\n"
            "Ad Creative Text: Restore your natural smile with our German-certified implantologists. Book your free scan.\n"
            "Observed CTA: Send WhatsApp Message\n"
            "Ad Count: 1\n"
            "DO NOT INVENT ad spend, ROAS, or conversion metrics. Return a structured AdsAnalysisResult."
        )

        response = await self.client.post(
            "/api/v1/run",
            json={
                "prompt": prompt,
                "agent_type": "ads_specialist",
                "timeout_seconds": 120.0
            }
        )
        self.assertEqual(response.status_code, 200)
        data = response.json()
        self.assertTrue(data["success"], f"Ads specialist failed: {data.get('error')}")

        output = data["output"]
        validated = AdsAnalysisResult.model_validate(output)
        self.assertEqual(validated.status, "active_ads")
        self.assertIsNone(validated.ad_spend)
        self.assertIsNone(validated.roas)
        self.assertIsNone(validated.conversion_rate)

    async def test_04_maps_specialist_live_execution_no_sentiment_fabrication(self):
        """Verify Maps Specialist analyzes Google Maps data and does NOT fabricate review sentiment."""
        prompt = (
            "Analyze Google Maps data for 'Apex Dental Spa':\n"
            "Business Name: Apex Dental Spa\n"
            "Category: Dental clinic\n"
            "Rating: 4.9\n"
            "Review Count: 140\n"
            "Address: DLF Phase 5, Gurgaon, Haryana\n"
            "Phone: +91 9876543210\n"
            "Website Present: True\n"
            "Instagram Present: True\n"
            "Review Text: NONE PROVIDED\n"
            "DO NOT FABRICATE review sentiment or customer complaints. Return a structured MapsAnalysisResult."
        )

        response = await self.client.post(
            "/api/v1/run",
            json={
                "prompt": prompt,
                "agent_type": "maps_specialist",
                "timeout_seconds": 120.0
            }
        )
        self.assertEqual(response.status_code, 200)
        data = response.json()
        self.assertTrue(data["success"], f"Maps specialist failed: {data.get('error')}")

        output = data["output"]
        validated = MapsAnalysisResult.model_validate(output)
        self.assertIn(validated.status, ["verified", "unverified"])
        self.assertEqual(validated.rating, 4.9)
        self.assertEqual(validated.review_count, 140)
        # Invariant: Review sentiment must be None or unknown
        self.assertIn(str(validated.review_sentiment).lower(), ["none", "unknown", ""])

    async def test_05_parallel_specialist_orchestration_endpoint(self):
        """Verify /api/v1/specialists/research executes triage and parallel specialists."""
        candidate_payload = {
            "business_name": "Apex Dental Spa",
            "business_type": "Cosmetic Dentistry",
            "source_platform": "google_maps",
            "source_url": "https://maps.google.com/apexdental",
            "scorecard_score": 9,
            "priority_tier": "immediate",
            "website_url": "https://apexdentalspa.in",
            "has_active_ads": True,
            "rating": 4.9,
            "review_count": 140,
            "address": "DLF Phase 5, Gurgaon",
            "phone_number": "+919876543210",
            "raw_metadata": {
                "category": "Dental clinic",
                "headline": "Dental Implants Special Offer",
                "ad_creative_text": "German implants with lifetime warranty",
                "cta_text": "Book Appointment"
            }
        }

        # Use pre-determined triage to verify specialist parallel execution directly
        pre_triage = {
            "qualified": True,
            "priority": "high",
            "reason": "High-value dental clinic in Gurgaon with active ads",
            "research_agents": ["ads", "maps"]
        }

        response = await self.client.post(
            "/api/v1/specialists/research",
            json={
                "candidate": candidate_payload,
                "triage_result": pre_triage,
                "limits": {"max_deep_research_leads": 5}
            }
        )
        self.assertEqual(response.status_code, 200)
        data = response.json()

        validated = SpecialistResearchAggregate.model_validate(data)
        self.assertEqual(validated.business_name, "Apex Dental Spa")
        self.assertIsNotNone(validated.ads_analysis)
        self.assertIsNotNone(validated.maps_analysis)
        self.assertIn("ads", validated.execution_status)
        self.assertIn("maps", validated.execution_status)

    async def test_06_credential_safety_invariant(self):
        """Verify API keys and sensitive tokens are never exposed in specialist responses."""
        response = await self.client.post(
            "/api/v1/run",
            json={
                "prompt": "Evaluate candidate 'Test Clinic' with rating 4.5.",
                "agent_type": "lead_triage",
                "timeout_seconds": 60.0
            }
        )
        body_text = response.text
        api_key = self.settings.AI_API_KEY.get_secret_value() if self.settings.AI_API_KEY else ""
        if api_key and len(api_key) > 5:
            self.assertNotIn(api_key, body_text)
        apify_token = os.getenv("APIFY_API_TOKEN", "")
        if apify_token and len(apify_token) > 5:
            self.assertNotIn(apify_token, body_text)


if __name__ == "__main__":
    unittest.main()
